test/unit/atomic_iteration_test.lua

```diff
@@ -0,0 +1,120 @@
+local expirationd = require("expirationd")
+local t = require("luatest")
+local g = t.group("atomic_iteration")
+
+local helpers = require("test.helper")
+
+g.before_all(function()
+    helpers.init_spaces(g)
+end)
+
+g.after_each(function()
+    helpers.truncate_spaces(g)
+end)
+
+function g.test_passing()
+    local task = expirationd.start("clean_all", g.tree.id, helpers.is_expired_true)
+    t.assert_equals(task.atomic_iteration, false)
+    task:kill()
+
+    task = expirationd.start("clean_all", g.tree.id, helpers.is_expired_true,
+            {atomic_iteration = true})
+    t.assert_equals(task.atomic_iteration, true)
+    task:kill()
+
+    --errors
+    t.assert_error_msg_content_equals("Invalid type of atomic_iteration, expected boolean",
+            expirationd.start, "clean_all", g.tree.id, helpers.is_expired_true,
+            {atomic_iteration = 1})
+end
+
+function g.test_memtx()
+    helpers.iteration_result = {}
+
+    local transactions = {}
+    function f(iterator)
+        local transaction = {}
+        -- old / new_tuple is not passed for vinyl
+        for request_number, old_tuple, new_tuple, space_id in iterator() do
+            table.insert(transaction, old_tuple:totable())
+        end
+        table.insert(transactions, transaction)
+    end
+
+    local true_box_begin = box.begin
+
+    -- mock box.begin
+    box.begin = function ()
+        true_box_begin()
+        box.on_commit(f)
+    end
+
+    for _, space in pairs({g.tree, g.hash}) do
+        -- tuples expired in one atomic_iteration
+        space:insert({1, "3"})
+        space:insert({2, "2"})
+        space:insert({3, "1"})
+
+
+        local task = expirationd.start("clean_all", space.id, helpers.is_expired_debug,
+                {atomic_iteration = true})
+        -- wait for tuples expired
+        helpers.retrying({}, function()
+            if space.index[0].type == "HASH" then
+                t.assert_equals(helpers.iteration_result, {{3, "1"}, {2, "2"}, {1, "3"}})
+            else
+                t.assert_equals(helpers.iteration_result, {{1, "3"}, {2, "2"}, {3, "1"}})
+            end
+        end)
+        task:kill()
```

Comment:
What happens if you try to kill a fiber during open transaction?

Exception on the first yield? It's also relevant for vinyl. I don't know about memtx to be honest. Maybe with enabled mvcc opened transaction will hang in memory (I suggest you to verify or refute my concerns).

Reply:
There should also be tarantool/tarantool tests for this

Reply:
No. It's your responsibility to close transaction.

Reply:
I mean that it would be great to introduce some flag that fiber will check on each iteration. E.g. is_finished. After "is_finished" is changed to true fiber should gracefully shutdown its work.

Reply:
Was added worker_canceled flag for that